core/core.go

```diff
@@ -0,0 +1,125 @@
+package core
+
+import (
+	"errors"
+	"io"
+	"time"
+)
+
+// Errors
+var (
+	ErrProcessNotFinished     = errors.New("Error: Process not finished")
+	ErrProcessNotStarted      = errors.New("Error: process not started yet")
+	ErrProcessAlreadyFinished = errors.New("Error: process already finished")
+	ErrProcessAlreadyStarted  = errors.New("Error: process already started")
+)
+
+// AppBus signals
+var (
+	SignalBuildComplete = `/build/complete/(?P<token>\w+)$`
+	SignalBuildStarted  = `/build/started/(?P<token>\w+)$`
+)
+
+type (
+	// App is defined by $NGBUILD_WORKSPACE/apps/$appname/config.yaml
+	// Apps will define different builds, for different projects or different kinds of builds or whatever
+	// Config has $NGBUILD_WORKSPACE/config.yaml as a parent and then applies
+	// $NGBUILD_WORKSPACE/apps/$appname/config.yaml onto it
+	// Everything App should be thread safe as it will be called from many goroutines
+	App interface {
+		Name() string
+		Config(namespace string, conf interface{}) error
+
+		// SendEvent is a dispatcher, it will send this string across all the apps integrations and also Core
+		SendEvent(event string)
+
+		// Listen will provide a channel to select on for a given regular expression
+		// returned map is the captured groups and values
+		Listen(event string) <-chan map[string]string
+
+		// NewBuild will be used by github and the like to create new builds for this app whenever they deem so
+		NewBuild(group string, config *BuildConfig) (token string, err error)
+
+		GetBuild(token string) (Build, error)
+	}
+
+	// BuildConfig describes a build, heavily in favour of github/git at the moment
+	BuildConfig struct {
```

Comment:
hrm, makes me think about other kinds of shared information, there is probably lots of stuff that for example slack integration wants to know that the github integration would need to provide, can't think of anything i'm happy with though, best of a bad bunch would be just storing explicitly named things in the metadata and doing 
```go
package slack

import (
    "github.com/watchly/ngbuild/integrations/github"
)

...
    buildDescription, ok := build.metadata[github.MetadataDescription]
...
```